Add codemod for migrating `@variants` and `@responsive` directives

RobinMalfait opened the pull request:
This PR migrates the `@variants` and `@responsive` directives.

In Tailwind CSS v2, these were used to generate certain variants of responsive variants for the give classes. In Tailwind CSS v3, these still worked but were implemented as a no-op such that these directives don't end up in your final CSS.

In Tailwind CSS v4, these don't exist at all anymore, so we can safely get rid of them by replacing them with their contents.

Input:
```css
@variants hover, focus {
  .foo {
    color: red;
  }
}

@responsive {
  .bar {
    color: blue;
  }
}
```

Output:
```css
.foo {
  color: red;
}

.bar {
  color: blue;
}
```

RobinMalfait commented:
* **#14749** <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14749?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#14748** <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14748?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `next`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @RobinMalfait and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

RobinMalfait commented:
The confusing part here is that if you use `@variants hover`
```css
@variants hover {
  .foo {
    color: red;
  }
}
```

If you then use it in your html as:
```html
<div class="hover:foo focus:foo foo">hello world</div>
```

Then the following CSS is being generated:
```css
.foo {
  color: red;
}
.hover\:foo:hover {
  color: red;
}
.focus\:foo:focus {
  color: red;
}
```

Which means that the `variants` listed in `@variants` don't matter. But `foo` is treated as a utility. So maybe the better solution is to wrap this in `@layer utilities` and migrate it to an `@utility` instead 🤔

philipp-spiess commented:
@RobinMalfait Ah, yeah, I think you're right we should migrate this to `@utility` to not break stuff. Fun that it doesn't matter which variants you used anymore but makes it easier for us at least 🙈

RobinMalfait commented:
Haha yeah that is a bit confusing. In v2 we generated everything ahead-of-time, and later introduced the just-in-time in v3. So in v2 we needed that information to generate all the CSS.

thecrypticace commented on packages/@tailwindcss-upgrade/src/codemods/migrate-tailwind-directives.test.ts:
```diff
@@ -400,3 +400,37 @@ it('should drop `@tailwind variants;`', async () => {
+      .foo {
+        color: red;
+      }
+    }"
```
Either the test name is wrong or the snapshot is wrong

RobinMalfait replied:
Deleted it now; This was moved to a separate file. Should be all good now.

philipp-spiess reviewed (changes requested):
Need to add the `migrateVariantsDirective` rule to the defaults now 🙈  Let's add a full integration test to make sure it indeed works

RobinMalfait commented:
Should be good to go now

philipp-spiess commented:
### Merge activity

* **Oct 22, 12:07 PM EDT**: The merge label 'merge-queue' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=tailwindlabs&repo=tailwindcss) once it meets the requirements.

RobinMalfait committed:
Add codemod for migrating `@variants` and `@responsive` directives (#14748)

This PR migrates the `@variants` and `@responsive` directives.

In Tailwind CSS v2, these were used to generate certain variants of responsive variants for the give classes. In Tailwind CSS v3, these still worked but were implemented as a no-op such that these directives don't end up in your final CSS.

In Tailwind CSS v4, these don't exist at all anymore, so we can safely get rid of them by replacing them with their contents.

Input:
```css
@variants hover, focus {
  .foo {
    color: red;
  }
}

@responsive {
  .bar {
    color: blue;
  }
}
```

Output:
```css
.foo {
  color: red;
}

.bar {
  color: blue;
}
```
